core/src/main/scala/org/http4s/Query.scala

```diff
@@ -1,13 +1,82 @@
 package org.http4s
 
-import org.http4s.Query._
+import org.http4s.FormQuery._
 import org.http4s.parser.QueryParser
-import org.http4s.util.{UrlFormCodec, UrlCodingUtils, Writer, Renderable}
+import org.http4s.util._
+import org.http4s.util.encoding.UriCodingUtils
 
 import scala.collection.generic.CanBuildFrom
-import scala.collection.immutable.{BitSet, IndexedSeq}
+import scala.collection.immutable.IndexedSeq
 import scala.collection.mutable.ListBuffer
-import scala.collection.{ IndexedSeqOptimized, mutable }
+import scala.collection.{IndexedSeqOptimized, mutable}
+
+
+sealed trait Query extends QueryOps with Renderable {
+  def asForm: FormQuery
+  def isEmpty: Boolean
+  def encoded: String
+}
+
+/** Represents the absence of a query component */
+object EmptyQuery extends Query {
+  override def asForm: FormQuery = FormQuery()
+  override def encoded: String = "<please check isEmpty before using this value>"
```

Comment:
so we want it on `PlainQuery` (<-- open to alternative names) and `FormQuery`, but I think that would mean we want `.asForm` and `.asPlain` on `Query`, but I don't think `FormQuery` can implement `.asPlain`, which means `Query` can't have `.asPlain`, or we still end up with an exceptional case.

Alternatively, we get rid of `Query` and use `FormQuery` or `PlainQuery` directly (freeing up the name `Query`, btw), but then which one goes into `Uri`?  Or else we have `Uri[PlainQuery]` :disappointed: